builtin/gc.c

```diff
@@ -1347,6 +1350,128 @@ static int maintenance_task_incremental_repack(struct maintenance_run_opts *opts
 	return 0;
 }
 
+static void link_or_copy_or_die(const char *src, const char *dst)
+{
+	if (!link(src, dst))
+		return;
+
+	warning_errno(_("failed to link '%s' to '%s'"), src, dst);
+
+	if (!copy_file(dst, src, 0644))
+		return;
+
+	die_errno(_("failed to copy '%s' to '%s'"), src, dst);
+}
+
+static void migrate_pack(const char *srcdir, const char *dstdir,
+			 const char *pack_filename)
+{
+	struct stat st;
+	char *basename;
+	struct strbuf src = STRBUF_INIT, dst = STRBUF_INIT;
+	struct {
+		const char *ext;
+		unsigned move:1;
+	} files[] = {
+		{".pack", 0},
+		{".keep", 0},
+		{".rev", 0},
+		{".idx", 1}, /* The index file must be atomically moved last. */
+	};
+
+	trace2_region_enter("maintenance", "migrate_pack", the_repository);
+
+	basename = xstrndup(pack_filename, strlen(pack_filename) - 5 /*.pack*/);
+
+	/* A pack without an index file is not yet ready to be migrated. */
+	strbuf_addf(&src, "%s/%s%s", srcdir, basename, ".idx");
+	if (stat(src.buf, &st))
```

Comment:
It would probably be more idiomatic to write `file_exists(src.buf)` here, but then, it would not buy us much because that function does the exact same thing (without bothering to verify that the entity is actually a _file_). So: 🤷